[SPARK-17729] [SQL] Enable creating hive bucketed tables

tejasapatil opened the pull request:
## What changes were proposed in this pull request?

Hive allows inserting data to bucketed table without guaranteeing bucketed and sorted-ness based on these two configs : `hive.enforce.bucketing` and `hive.enforce.sorting`.

What does this PR achieve ?
- Spark will disallow users from writing outputs to hive bucketed tables by default (given that output won't adhere with Hive's semantics).
- IF user still wants to write to hive bucketed table, the only resort is to use `hive.enforce.bucketing=false` and `hive.enforce.sorting=false` which means user does NOT care about bucketing guarantees.

Changes done in this PR:
- Extract table's bucketing information in `HiveClientImpl`
- While writing table info to metastore, `HiveClientImpl` now populates the bucketing information in the hive `Table` object
- `InsertIntoHiveTable` allows inserts to bucketed table only if both `hive.enforce.bucketing` and `hive.enforce.sorting` are `false`

Ability to create bucketed tables will enable adding test cases to Spark while I add more changes related to hive bucketing support. Design doc for hive hive bucketing support : https://docs.google.com/document/d/1a8IDh23RAkrkg9YYAeO51F4aGO8-xAlupKwdshve2fc/edit#

## How was this patch tested?
- Added test for creating bucketed and sorted table.
- Added test to ensure that INSERTs fail if strict bucket / sort is enforced
- Added test to ensure that INSERTs can go through if strict bucket / sort is NOT enforced
- Added test to validate that bucketing information shows up in output of DESC FORMATTED
- Added test to ensure that `SHOW CREATE TABLE` works for hive bucketed tables

tejasapatil committed:
[SPARK-17729] [SQL] Enable creating hive bucketed tables

SparkQA commented:
**[Test build #75829 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75829/testReport)** for PR 17644 at commit [`0348a96`](https://github.com/apache/spark/commit/0348a969a43872da484dd13f1f211b966839baec).

SparkQA commented:
**[Test build #75829 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75829/testReport)** for PR 17644 at commit [`0348a96`](https://github.com/apache/spark/commit/0348a969a43872da484dd13f1f211b966839baec).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75829/
Test FAILed.

tejasapatil committed:
fixing failed test cases : HiveDDLSuite and HiveExternalCatalogSuite

SparkQA commented:
**[Test build #75838 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75838/testReport)** for PR 17644 at commit [`fc04991`](https://github.com/apache/spark/commit/fc04991e288167fe76d821ca71b0a91294e241f0).

SparkQA commented:
**[Test build #75838 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75838/testReport)** for PR 17644 at commit [`fc04991`](https://github.com/apache/spark/commit/fc04991e288167fe76d821ca71b0a91294e241f0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75838/
Test PASSed.

tejasapatil commented:
cc @cloud-fan @hvanhovell @sameeragarwal for review

cloud-fan commented:
I'll review it after branch 2.2 is cut

SparkQA commented:
**[Test build #76298 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76298/testReport)** for PR 17644 at commit [`1b78141`](https://github.com/apache/spark/commit/1b7814123b19bc14e5f692ccb80bf9694c76e2e0).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76298/
Test FAILed.

tejasapatil commented:
Jenkins test this please

SparkQA commented:
**[Test build #76312 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76312/testReport)** for PR 17644 at commit [`1b78141`](https://github.com/apache/spark/commit/1b7814123b19bc14e5f692ccb80bf9694c76e2e0).

SparkQA commented:
**[Test build #76312 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76312/testReport)** for PR 17644 at commit [`1b78141`](https://github.com/apache/spark/commit/1b7814123b19bc14e5f692ccb80bf9694c76e2e0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76312/
Test PASSed.

tejasapatil commented:
@cloud-fan : ping !!

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:
```diff
@@ -632,9 +632,51 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
+      newProperties.put(property, value)
+    }
+
+    // Since the table schema will be changed, check if the bucketing and sort columns are still
```
I don't think we need to do this, according to the document of `ExternalCatalog.alterTableSchema`, the caller will guarantee that the new schema still contains partition columns and bucket columns.

tejasapatil replied:
I have to do this because one of the unit tests was failing : https://github.com/apache/spark/blob/703c42c398fefd3f7f60e1c503c4df50251f8dcf/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala#L246

In that test case, the table has a bucketing and alter removes the bucketing column. The way I interpret things: if the alter command is not removing the bucketing / sorting columns, we should keep the original bucketing spec. If any one of the bucketing / sorting columns are removed in alter, then the bucketing spec needs to be reset to empty (else we might refer to dangling columns).

cloud-fan replied:
does hive allow this? i.e. alter schema without bucket/sort columns.

tejasapatil replied:
Unfortunately YES. Below I alter a bucketed table to remove the bucketing column from its schema. Modified table still points to old bucket spec. This is bad.

```
hive> desc formatted bucketed_partitioned_1;

# col_name            	data_type           	comment
user_id             	bigint
name                	string

# Storage Information
....
Num Buckets:        	8
Bucket Columns:     	[user_id]
Sort Columns:       	[Order(col:user_id, order:1)]

hive> ALTER TABLE bucketed_partitioned_1 REPLACE COLUMNS ( eid INT, name STRING);
OK

hive> desc formatted bucketed_partitioned_1;
# col_name            	data_type           	comment
eid                 	int
name                	string

# Storage Information
.....
Num Buckets:        	8
Bucket Columns:     	[user_id]
Sort Columns:       	[Order(col:user_id, order:1)]
```

cloud-fan replied:
do you mean hive has a wrong behavior about this? Maybe we can diverge from hive and disallow altering schema without bucket/sort columns.

tejasapatil replied:
Yeah hive behavior is absurd. If the alter table statement removes bucketing columns, we should either disallow such operation (seems harsh) OR remove the bucketing spec (which is what I am doing in this PR).

>> ... disallow altering schema without bucket/sort columns.

I am not sure if I understand this right. Does the current approach in this PR match this or you are proposing something else ?

cloud-fan replied:
let's disallow removing bucketing columns in `SessionCatalog.alterTableSchema`, to make the logic easier, we can change the behavior later if we wanna support this.

tejasapatil replied:
Updated the PR with this change

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:
```diff
@@ -171,8 +172,7 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
             dataSchema = dataSchema,
-            // We don't support hive bucketed tables, only ones we write out.
-            bucketSpec = None,
+            bucketSpec = bucketSpec,
```
is it safe to do so? I think this PR only covers the write path?

tejasapatil replied:
I dont think this could cause problems (you can correct me if I am missing anything). Even if bucketing spec is set over here, this PR does not make use of that information in scan operators.

cloud-fan replied:
here we are reading hive table as a data source table, so the scan operator is actually `FileSourceScanExec`, which recognizes bucketing.

tejasapatil replied:
You are right. I have corrected this. Thanks for pointing this out !!

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -374,10 +375,39 @@ private[hive] class HiveClientImpl(
+          sortColumnOrders.map { sortOrder =>
+            val columnName = sortOrder.getCol
+
+            if (!cols.exists(_.name.equalsIgnoreCase(columnName))) {
```
why do we need to check this? Doesn't hive metastore already guarantees this?

tejasapatil replied:
I had added this for sanity. As you mentioned, hive metastore already guarantees this. I will remove this.

cloud-fan commented:
A high-level question: if we propagate the bucketing information of a hive table, can we correctly read this table with Spark SQL? Shall we have a config to change Spark's shuffle hash function to use hive hash?

tejasapatil commented:
@cloud-fan : Thanks for reviewing !!

> if we propagate the bucketing information of a hive table, can we correctly read this table with Spark SQL?

With this PR, `HiveTableScanExec` still doesn't advertise the table to be bucketed so optimizer wont treat the data as bucketed. tldr: It would be read just like a regular table (as it happens now in trunk).

> Shall we have a config to change Spark's shuffle hash function to use hive hash?

Yeah. We need to add that but I will do that in a separate PR. I had a PR from last year but given that there are many changes in SQL land, I couldn't rebase it. Am working on creating a new PR instead after studying the new changes since last year.

tejasapatil committed:
review comment #1

SparkQA commented:
**[Test build #76460 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76460/testReport)** for PR 17644 at commit [`e3c41bf`](https://github.com/apache/spark/commit/e3c41bfef40cb50710ef7f573a7ce71e16f739d5).

tejasapatil commented:
Jenkins test this please

SparkQA commented:
**[Test build #76462 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76462/testReport)** for PR 17644 at commit [`6e6e767`](https://github.com/apache/spark/commit/6e6e767c9a6787965d6eb9a32608aacacd543e23).

SparkQA commented:
**[Test build #76462 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76462/testReport)** for PR 17644 at commit [`6e6e767`](https://github.com/apache/spark/commit/6e6e767c9a6787965d6eb9a32608aacacd543e23).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76462/
Test PASSed.

SparkQA commented:
**[Test build #76460 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76460/testReport)** for PR 17644 at commit [`e3c41bf`](https://github.com/apache/spark/commit/e3c41bfef40cb50710ef7f573a7ce71e16f739d5).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76460/
Test PASSed.

tejasapatil commented:
Jenkins test this please

SparkQA commented:
**[Test build #76506 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76506/testReport)** for PR 17644 at commit [`6e6e767`](https://github.com/apache/spark/commit/6e6e767c9a6787965d6eb9a32608aacacd543e23).

SparkQA commented:
**[Test build #76506 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76506/testReport)** for PR 17644 at commit [`6e6e767`](https://github.com/apache/spark/commit/6e6e767c9a6787965d6eb9a32608aacacd543e23).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76506/
Test FAILed.

tejasapatil commented:
Jenkins test this please

SparkQA commented:
**[Test build #76515 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76515/testReport)** for PR 17644 at commit [`6e6e767`](https://github.com/apache/spark/commit/6e6e767c9a6787965d6eb9a32608aacacd543e23).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76515/
Test FAILed.

tejasapatil commented:
Jenkins test this please

SparkQA commented:
**[Test build #76529 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76529/testReport)** for PR 17644 at commit [`6e6e767`](https://github.com/apache/spark/commit/6e6e767c9a6787965d6eb9a32608aacacd543e23).

SparkQA commented:
**[Test build #76529 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76529/testReport)** for PR 17644 at commit [`6e6e767`](https://github.com/apache/spark/commit/6e6e767c9a6787965d6eb9a32608aacacd543e23).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76529/
Test PASSed.

SparkQA commented:
**[Test build #76588 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76588/testReport)** for PR 17644 at commit [`ab12943`](https://github.com/apache/spark/commit/ab12943c724a4549616bc19c612f4dc9ebdfa05f).

SparkQA commented:
**[Test build #76588 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76588/testReport)** for PR 17644 at commit [`ab12943`](https://github.com/apache/spark/commit/ab12943c724a4549616bc19c612f4dc9ebdfa05f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76588/
Test FAILed.

tejasapatil commented:
Jenkins test this please

SparkQA commented:
**[Test build #76595 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76595/testReport)** for PR 17644 at commit [`7a0be86`](https://github.com/apache/spark/commit/7a0be86d11f274d00ff5cd6d67b4a89c75428cce).

SparkQA commented:
**[Test build #76595 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76595/testReport)** for PR 17644 at commit [`7a0be86`](https://github.com/apache/spark/commit/7a0be86d11f274d00ff5cd6d67b4a89c75428cce).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76595/
Test FAILed.

SparkQA commented:
**[Test build #76613 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76613/testReport)** for PR 17644 at commit [`49040e8`](https://github.com/apache/spark/commit/49040e83217a787f7a995f9da941617885e10821).

SparkQA commented:
**[Test build #76613 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76613/testReport)** for PR 17644 at commit [`49040e8`](https://github.com/apache/spark/commit/49040e83217a787f7a995f9da941617885e10821).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76613/
Test FAILed.

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #76676 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76676/testReport)** for PR 17644 at commit [`49040e8`](https://github.com/apache/spark/commit/49040e83217a787f7a995f9da941617885e10821).

SparkQA commented:
**[Test build #76676 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76676/testReport)** for PR 17644 at commit [`49040e8`](https://github.com/apache/spark/commit/49040e83217a787f7a995f9da941617885e10821).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76676/
Test PASSed.

SparkQA commented:
**[Test build #76703 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76703/testReport)** for PR 17644 at commit [`8aaff4c`](https://github.com/apache/spark/commit/8aaff4c048af3466774719c93086cb8eb0357630).

SparkQA commented:
**[Test build #76703 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76703/testReport)** for PR 17644 at commit [`8aaff4c`](https://github.com/apache/spark/commit/8aaff4c048af3466774719c93086cb8eb0357630).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76703/
Test FAILed.

SparkQA commented:
**[Test build #76758 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76758/testReport)** for PR 17644 at commit [`c7b6260`](https://github.com/apache/spark/commit/c7b6260575fb417be276fb07dbe4bf2c1543bd32).

SparkQA commented:
**[Test build #76758 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76758/testReport)** for PR 17644 at commit [`c7b6260`](https://github.com/apache/spark/commit/c7b6260575fb417be276fb07dbe4bf2c1543bd32).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76758/
Test FAILed.

SparkQA commented:
**[Test build #76767 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76767/testReport)** for PR 17644 at commit [`fc9d94f`](https://github.com/apache/spark/commit/fc9d94ff1ed712418c2866c860cabc7827cf9b4e).

SparkQA commented:
**[Test build #76767 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76767/testReport)** for PR 17644 at commit [`fc9d94f`](https://github.com/apache/spark/commit/fc9d94ff1ed712418c2866c860cabc7827cf9b4e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76767/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala:
```diff
@@ -335,6 +336,32 @@ abstract class ExternalCatalog
     postToAll(RenameFunctionEvent(db, oldName, newName))
   }
 
+  final def checkSchemaContainsBucketingColumns(
```
we should move this check to `SessionCatalog.alterTableSchema`, which already contains some checks

tejasapatil replied:
But then `InMemoryCatalog.alterTableSchema` will not have this check. That makes few tests to fail. Dont we want to have consistency wrt this check in both `InMemoryCatalog` and  `SessionCatalog` ?

cloud-fan replied:
It will be very annoying to duplicate all checks in `SessionCatalog` and `ExternalCatalog`. I think we should put checks in `SessionCatalog` as it behaves like a front-end of `ExternalCatalog`

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:
```diff
@@ -902,9 +902,14 @@ case class ShowCreateTableCommand(table: TableIdentifier) extends RunnableComman
-    if (metadata.bucketSpec.isDefined) {
-      throw new UnsupportedOperationException(
-        "Creating Hive table with bucket spec is not supported yet.")
+    if (metadata.bucketSpec.nonEmpty) {
```
nit: `isDefined` is more ideal.

tejasapatil replied:
did this change

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -871,6 +886,23 @@ private[hive] object HiveClientImpl {
+    table.bucketSpec match {
+      case Some(bucketSpec) =>
+        hiveTable.setNumBuckets(bucketSpec.numBuckets)
+        hiveTable.setBucketCols(bucketSpec.bucketColumnNames.toList.asJava)
```
I think we can only do this for hive tables, not data source tables.

tejasapatil replied:
I didn't get you. Can you please elaborate ?

With this PR, here is the behavior:

bucketed table created using dataframe API

```
scala> df.write.format("org.apache.spark.sql.hive.orc.OrcFileFormat").bucketBy(8, "j", "k").sortBy("j", "k").saveAsTable("spark_native1")
17/05/11 07:34:23 WARN HiveExternalCatalog: Persisting bucketed data source table `default`.`spark_native1` into Hive metastore in Spark SQL specific format, which is NOT compatible with Hive.

scala> spark.sql(" desc formatted table1 ").collect.foreach(println)
[# col_name,data_type,comment]
[i,int,null]
[j,int,null]
[k,string,null]
[,,]
[# Detailed Table Information,,]
[Database,default,]
[Table,table1,]
[Owner,tejasp,]
[Created,Fri Apr 14 17:35:51 PDT 2017,]
[Last Access,Wed Dec 31 16:00:00 PST 1969,]
[Type,MANAGED,]
[Provider,org.apache.spark.sql.hive.orc.OrcFileFormat,]
[Num Buckets,8,]
[Bucket Columns,[`j`, `k`],]
[Sort Columns,[`j`, `k`],]
[Properties,[serialization.format=1],]
[Statistics,2938 bytes, 16 rows,]
[Location,file:/Users/tejasp/Desktop/dev/apache-hive-1.2.1-bin/warehouse/table1,]
[Serde Library,org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe,]
[InputFormat,org.apache.hadoop.mapred.SequenceFileInputFormat,]
[OutputFormat,org.apache.hadoop.hive.ql.io.HiveSequenceFileOutputFormat,]

# Hive CLI
hive>  desc formatted spark_native1 ;
OK
# col_name            	data_type           	comment

col                 	array<string>       	from deserializer

# Detailed Table Information
Database:           	default
Owner:              	tejasp
CreateTime:         	Thu May 11 07:34:23 PDT 2017
LastAccessTime:     	UNKNOWN
Protect Mode:       	None
Retention:          	0
Location:           	file:/Users/tejasp/Desktop/dev/apache-hive-1.2.1-bin/warehouse/spark_native1
Table Type:         	MANAGED_TABLE
Table Parameters:
	COLUMN_STATS_ACCURATE	false
	numFiles            	8
	numRows             	-1
	rawDataSize         	-1
	spark.sql.sources.provider	org.apache.spark.sql.hive.orc.OrcFileFormat
	spark.sql.sources.schema.bucketCol.0	j
	spark.sql.sources.schema.bucketCol.1	k
	spark.sql.sources.schema.numBucketCols	2
	spark.sql.sources.schema.numBuckets	8
	spark.sql.sources.schema.numParts	1
	spark.sql.sources.schema.numSortCols	2
	spark.sql.sources.schema.part.0	{\"type\":\"struct\",\"fields\":[{\"name\":\"i\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}},{\"name\":\"j\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}},{\"name\":\"k\",\"type\":\"string\",\"nullable\":true,\"metadata\":{}}]}
	spark.sql.sources.schema.sortCol.0	j
	spark.sql.sources.schema.sortCol.1	k
	totalSize           	2938
	transient_lastDdlTime	1494513263

# Storage Information
SerDe Library:      	org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe
InputFormat:        	org.apache.hadoop.mapred.SequenceFileInputFormat
OutputFormat:       	org.apache.hadoop.hive.ql.io.HiveSequenceFileOutputFormat
Compressed:         	No
Num Buckets:        	-1
Bucket Columns:     	[]
Sort Columns:       	[]
Storage Desc Params:
	path                	file:/Users/tejasp/Desktop/dev/apache-hive-1.2.1-bin/warehouse/spark_native1
	serialization.format	1
Time taken: 0.057 seconds, Fetched: 42 row(s)
```

bucketed table created via DDL

```
scala> hc.sql("""
     | CREATE TABLE `bucket_table_spark_created`(`user_id` int,`name` string)
     | PARTITIONED BY (`ds` string)
     | CLUSTERED BY (user_id)
     | SORTED BY (user_id ASC)
     | INTO 8 BUCKETS
     | """)
res2: org.apache.spark.sql.DataFrame = []

scala> hc.sql(""" desc formatted bucket_table_spark_created """).collect.foreach(println)
[# col_name,data_type,comment]
[user_id,int,null]
[name,string,null]
[ds,string,null]
[# Partition Information,,]
[# col_name,data_type,comment]
[ds,string,null]
[,,]
[# Detailed Table Information,,]
[Database,default,]
[Table,bucket_table_spark_created,]
[Owner,tejasp,]
[Created,Thu May 11 07:48:07 PDT 2017,]
[Last Access,Wed Dec 31 16:00:00 PST 1969,]
[Type,MANAGED,]
[Provider,hive,]
[Num Buckets,8,]
[Bucket Columns,[`user_id`],]
[Sort Columns,[`user_id`],]
[Properties,[serialization.format=1],]
[Location,file:/Users/tejasp/Desktop/dev/apache-hive-1.2.1-bin/warehouse/bucket_table_spark_created,]
[Serde Library,org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe,]
[InputFormat,org.apache.hadoop.mapred.TextInputFormat,]
[OutputFormat,org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat,]
[Partition Provider,Catalog,]


# Hive CLI
hive> desc formatted bucket_table_spark_created ;
OK
# col_name            	data_type           	comment

user_id             	int
name                	string

# Partition Information
# col_name            	data_type           	comment

ds                  	string

# Detailed Table Information
Database:           	default
Owner:              	tejasp
CreateTime:         	Thu May 11 07:48:07 PDT 2017
LastAccessTime:     	UNKNOWN
Protect Mode:       	None
Retention:          	0
Location:           	file:/Users/tejasp/Desktop/dev/apache-hive-1.2.1-bin/warehouse/bucket_table_spark_created
Table Type:         	MANAGED_TABLE
Table Parameters:
	spark.sql.sources.schema.bucketCol.0	user_id
	spark.sql.sources.schema.numBucketCols	1
	spark.sql.sources.schema.numBuckets	8
	spark.sql.sources.schema.numPartCols	1
	spark.sql.sources.schema.numParts	1
	spark.sql.sources.schema.numSortCols	1
	spark.sql.sources.schema.part.0	{\"type\":\"struct\",\"fields\":[{\"name\":\"user_id\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}},{\"name\":\"name\",\"type\":\"string\",\"nullable\":true,\"metadata\":{}},{\"name\":\"ds\",\"type\":\"string\",\"nullable\":true,\"metadata\":{}}]}
	spark.sql.sources.schema.partCol.0	ds
	spark.sql.sources.schema.sortCol.0	user_id
	transient_lastDdlTime	1494514087

# Storage Information
SerDe Library:      	org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe
InputFormat:        	org.apache.hadoop.mapred.TextInputFormat
OutputFormat:       	org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat
Compressed:         	No
Num Buckets:        	8
Bucket Columns:     	[user_id]
Sort Columns:       	[Order(col:user_id, order:1)]
Storage Desc Params:
	serialization.format	1
Time taken: 1.169 seconds, Fetched: 41 row(s)
```

cloud-fan replied:
For data source table, which can be created by `CREATE TABLE src(...) USING parquet ...`, the bucketing information is in table properties, and hive will always read this table as a non-bucketed table.

After your PR, for bucketed data source tables written by Spark, Hive will read them as bucketed tables and cause problems, because the bucket hashing function is different.

tejasapatil replied:
Got it. Adding a check if the table is hive and not data source table.

tejasapatil committed:
nonEmpty => isDefined

SparkQA commented:
**[Test build #76815 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76815/testReport)** for PR 17644 at commit [`d01c721`](https://github.com/apache/spark/commit/d01c721cb793ea8344ee3479652795662ca39968).

SparkQA commented:
**[Test build #76815 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76815/testReport)** for PR 17644 at commit [`d01c721`](https://github.com/apache/spark/commit/d01c721cb793ea8344ee3479652795662ca39968).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76815/
Test PASSed.

tejasapatil committed:
check explicitly for hive tables

SparkQA commented:
**[Test build #76852 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76852/testReport)** for PR 17644 at commit [`239beee`](https://github.com/apache/spark/commit/239beee4f74488445c4e60c5aa17f54004ce9658).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76852/
Test FAILed.

cloud-fan commented:
retest this please.

SparkQA commented:
**[Test build #76860 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76860/testReport)** for PR 17644 at commit [`239beee`](https://github.com/apache/spark/commit/239beee4f74488445c4e60c5aa17f54004ce9658).

SparkQA commented:
**[Test build #76860 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76860/testReport)** for PR 17644 at commit [`239beee`](https://github.com/apache/spark/commit/239beee4f74488445c4e60c5aa17f54004ce9658).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76860/
Test FAILed.

tejasapatil committed:
remove check as `alterTableSchema` ensures that dropping existing columns is not allowed

SparkQA commented:
**[Test build #76882 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76882/testReport)** for PR 17644 at commit [`10a9fd0`](https://github.com/apache/spark/commit/10a9fd0458911ea2bee082a8be58cf19ba870d2a).

SparkQA commented:
**[Test build #76882 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76882/testReport)** for PR 17644 at commit [`10a9fd0`](https://github.com/apache/spark/commit/10a9fd0458911ea2bee082a8be58cf19ba870d2a).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76882/
Test PASSed.

SparkQA commented:
**[Test build #76884 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76884/testReport)** for PR 17644 at commit [`d6ce8b5`](https://github.com/apache/spark/commit/d6ce8b500acd1d1c0f2d1e6a89de5c4e0266e4c4).

SparkQA commented:
**[Test build #76887 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76887/testReport)** for PR 17644 at commit [`0aa8539`](https://github.com/apache/spark/commit/0aa853940b9a79db059ae128c3f9f7d989312d60).

SparkQA commented:
**[Test build #76884 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76884/testReport)** for PR 17644 at commit [`d6ce8b5`](https://github.com/apache/spark/commit/d6ce8b500acd1d1c0f2d1e6a89de5c4e0266e4c4).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76884/
Test PASSed.

SparkQA commented:
**[Test build #76887 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76887/testReport)** for PR 17644 at commit [`0aa8539`](https://github.com/apache/spark/commit/0aa853940b9a79db059ae128c3f9f7d989312d60).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76887/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala:
```diff
@@ -17,6 +17,7 @@
 
 package org.apache.spark.sql.catalyst.catalog
 
+import org.apache.spark.sql.AnalysisException
```
please remove these unnecessary changes in this PR.

tejasapatil replied:
done

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -408,9 +425,7 @@ private[hive] class HiveClientImpl(
-        // We can not populate bucketing information for Hive tables as Spark SQL has a different
-        // implementation of hash function from Hive.
-        bucketSpec = None,
+        bucketSpec = bucketSpec,
```
please add a comment to say that, for data source tables, we will always overwrite the bucket spec in `HiveExternalCatalog` with the bucketing information in table properties.

tejasapatil replied:
done

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala:
```diff
@@ -307,6 +307,27 @@ case class InsertIntoHiveTable(
+          throw new AnalysisException(message)
+        } else {
+          logWarning(message + s" Inserting data anyways since both $enforceBucketingConfig and " +
+            s"$enforceSortingConfig are set to false.")
```
shall we remove the bucket properties of the table in this case? what does hive do?

tejasapatil replied:
- In Hive 1.x:
  - If enforcing configs are set: data is populated respecting the tables' bucketing spec
  - If enforcing configs are NOT set: data is populated and not conformant with the tables' bucketing spec
- In Hive 2.x, these configs are not there and hive would always populate data conformant with table's bucketing.

With Spark, currently the data would be written out in a non conformant way despite of that config being set or not. This PR will go to the model of Hive 1.x. I am working on a next PR which would make things like Hive 2.0.

cloud-fan replied:
so after insertion(if not enforcing), the table is still a buckted table but read it will cause wrong result?

tejasapatil replied:
In hive: It would lead to wrong result.

In spark (over master and also after this PR): the table scan operation does not take bucketing into account so it would be read as a regular table. So, it won't be read "wrong", its just that we wont take advantage of bucketing.

tejasapatil committed:
review comment

SparkQA commented:
**[Test build #76909 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76909/testReport)** for PR 17644 at commit [`efc8c8b`](https://github.com/apache/spark/commit/efc8c8bd9963e576004bae71cb144788db00eab5).

SparkQA commented:
**[Test build #76910 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76910/testReport)** for PR 17644 at commit [`bf306da`](https://github.com/apache/spark/commit/bf306da2ea7d87887205bdc875ce849e897bb233).

SparkQA commented:
**[Test build #76909 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76909/testReport)** for PR 17644 at commit [`efc8c8b`](https://github.com/apache/spark/commit/efc8c8bd9963e576004bae71cb144788db00eab5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76909/
Test PASSed.

SparkQA commented:
**[Test build #76910 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76910/testReport)** for PR 17644 at commit [`bf306da`](https://github.com/apache/spark/commit/bf306da2ea7d87887205bdc875ce849e897bb233).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76910/
Test PASSed.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -408,9 +425,9 @@ private[hive] class HiveClientImpl(
-        // We can not populate bucketing information for Hive tables as Spark SQL has a different
-        // implementation of hash function from Hive.
-        bucketSpec = None,
+        // For data source tables, we will always overwrite the bucket spec in
```
Sorry I made a mistake. This is true even for hive tables.

If the table is written by Spark, we will put bucketing information in table properties, and will always overwrite the bucket spec in hive metastore by the bucketing information in table properties. This means, if we have bucket spec in both hive metastore and table properties, we will trust the one in table properties.

let's update this document.

tejasapatil replied:
updated

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -870,6 +887,23 @@ private[hive] object HiveClientImpl {
     }
+
+    table.bucketSpec match {
+      case Some(bucketSpec) if DDLUtils.isHiveTable(table) =>
```
After thinking it more, I think this one is less important.

Our main goal is to allow spark to read bucketed tables written by hive, but not to allow hive to read bucketed tables written by spark.

How about we remove it for now and add it later with more discussion?

tejasapatil replied:
> but not to allow hive to read bucketed tables written by spark.

If Hive is NOT able to read the datasource tables which are bucketed, thats OK as its not compatible with hive. But for hive native tables, the interoperability amongst spark and hive is what I want.

cloud-fan replied:
ok makes sense

cloud-fan commented:
LGTM except 2 comments

tejasapatil committed:
fix documentation

SparkQA commented:
**[Test build #76941 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76941/testReport)** for PR 17644 at commit [`865711f`](https://github.com/apache/spark/commit/865711f87259245292c78098b3ce5381a99ffc55).

SparkQA commented:
**[Test build #76941 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76941/testReport)** for PR 17644 at commit [`865711f`](https://github.com/apache/spark/commit/865711f87259245292c78098b3ce5381a99ffc55).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76941/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala:
```diff
@@ -247,7 +247,7 @@ abstract class ExternalCatalogSuite extends SparkFunSuite with BeforeAndAfterEac
     val tbl1 = catalog.getTable("db2", "tbl1")
     val newSchema = StructType(Seq(
-      StructField("new_field_1", IntegerType),
+      StructField("col1", IntegerType),
```
@tejasapatil do you still remember why we update it?

tejasapatil replied:
yes. This was done because before this PR the test case was removing bucketed columns in the alter operation. We decided to disallow removing bucketing columns and support this if needed in future. 

Here is the discussion that we both had about this : https://github.com/apache/spark/pull/17644#discussion_r114784136

tejasapatil replied:
Huh. If I revert the change to this test case, it does not fail anymore. This is bad because the table properties still say that its bucketed over `col1` but `col1` is not in the modified table schema. I am taking a look to see what changed.
